pantheon-bundle/src/main/java/com/redhat/pantheon/model/api/Child.java

```diff
@@ -37,28 +41,92 @@
      * Indicates if the child exists.
      * @return True, if the child exists. False otherwise.
      */
-    boolean isPresent();
+    default boolean isPresent() {
+        return get() != null;
+    }
 
     /**
-     * Provides a null-safe way to operate on the value of the child, and return an
-     * {@link Optional} with the result of the operation. This allowes the caller to
-     * continue to operate in a null-safe fashion.
-     * @param func The function to apply to the value
+     * Convert this Child to an {@link Optional}
+     * @return An {@link Optional} with the contained value.
+     */
+    default Optional<T> asOptional() {
+        return Optional.ofNullable(get());
+    }
+
+    /**
+     * Navigates to the {@link Child} provided by an accessor function.
+     * @param childAccessor A function which given a {@link SlingModel} type, will
+     *                      yield a child
      * @param <R>
-     * @return An optional indicating the result of the operation. If the operation
-     * returns null, or if the value of this child was not present in the first place,
-     * this returns an empty Optional
-     * @deprecated Use {@link ResourceTraversal#traverseFrom(SlingModel)}
-     * for safe resource traversals
+     * @return A {@link Child} (may be non-existent) as indicated by the accessor.
      */
-    @Deprecated
-    <R> Optional<R> map(Function<? super T, ? extends R> func);
+    default <R extends SlingModel> Child<R> toChild(Function<? super T, Child<R>> childAccessor) {
+        if(isPresent()) {
+            return childAccessor.apply(get());
+        }
+        return (Child<R>) NullObjects.nullChild();
+    }
```

Comment:
I think this warrants some further explanation. It's easy to think that the `isPresent()` call applies to the child, when I think this must apply to the parent.

I realized this when I was trying to get to the bottom of why your changes help to make things null-safe. For example, if you have this chain:
```
Child.from(A)
.toChild(B)
.toChild(C)
.toChild(D)
.toChild(E)
.getAsOptional()
````
...and let's say that you get a null on C, then what the heck stops us from getting an NPE on the call to D?

I guess the answer must be that there's some code that makes sure that once C becomes a NullChild, then the call to get D must also return a NullChild, and the same for the call down into E. The code that I have attached this comment to must be what does that. But like I said, it's easy to misread at first glance, so if I'm right then I'd like to request *something* that makes this look less innocuous and more critically important (which I believe it is). Maybe some enhanced javadoc, or some `// INLINE COMMENTS IN ALL CAPS` or something - whatever you think would be most fitting.

Reply:
You are correct in your interpretation. This code is applying the Null Object pattern (https://en.wikipedia.org/wiki/Null_object_pattern) to ensure the null-safety. So let me make a few points which may clarify your questions:

1. In line 64 above, the call to `isPresent()` is done on the parent (aka the current node in the navigation). Perhaps it would be more self-explanatory to replace that with `this.isPresent()`.
2. Line 67 above is where the "magic" begins. If the current node is **not present**, then any child will inevitably not exist either, and so a `NullChild` is returned every time. 
3. The code never throws an NPE because it **never** returns a `null`, instead it returns a `NullChild` which has some specific behaviours to represent the fact that it is an empty node.

Happy to add more context and javadocs, just let me know where so I can do it in the right places where things were not clear. It's hard to identify those when you're the author.